rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,359 @@
+<Button>
+  <Loader as="div" />
+</Button>
+
+// Result
+<Button>
+  {/* 👇 this is just a plain div, Loader is not rendered 🤨 */}
+  <div />
+</Button>
+```
+
+In the above example `as` is intended to represent the `Loader` prop, not to define the component to be rendered. This is a result of special handling of the `as` prop which will always be omitted from the slot props and used as the component to be rendered.
+
+### Complex typings and unsafe internal implementation
+
+[microsoft/fluentui#17508](https://github.com/microsoft/fluentui/pull/17508/files) includes type fixes for `mergeProps` which provides extra type safety to validate shorthand slots.
+
+However, the result still leads to complicated component state types benefiting from the stricter typing through generics.
+
+### Native attribute filtering needs to happen by default
+
+For slots that render native elements, it's difficult to provide type safety in order to avoid unsafe DOM attribute spreads as shown in the below example:
+
+```tsx
+// ComponentState
+const state = {
+  buttonSlot: { as: 'button', type: 'submit' },
+};
+
+function App() {
+  // Typescript cannot warn that this is potentially evil
+  return <Component buttonSlot={{ as: 'div' }} />;
+}
+```
+
+```html
+<!-- Thankfully still works, but props must be filtered on every render -->
+<div></div>
+```
+
+The above example still works thanks to the the native attribute filtering that is done as part of `getSlots`. However, these filters always need to be run when resolving shorthand slots. During each render, every prop of a component is iterated and re-filtered.
+
+The current filtering also needs to iterate all the allowed props during a nested loop, although this could be optimized quite easily.
+
+While the existence of these helpers is necessary (as explained in the below proposals), it seems unnecessary that it needs to happen on every render.
+
+### Extending a component by adding/modifying slots
+
+One common use case as seen in `Menu`, `Avatar` and `Badge` is to extend a base component by adding or modifying a slot. There are two ways where it is possible to do this currently:
+
+```tsx
+// Extend by modifying existing slots
+// call resolveShorthandprops on defaultProps
+const state = mergeProps(
+  {
+    as: 'span',
+    label: { as: 'span' },
+    getInitials: defaultGetInitials,
+    ref,
+  },
+  defaultProps && resolveShorthandProps(defaultProps, avatarShorthandProps),
+  resolveShorthandProps(props, avatarShorthandProps),
+);
+```
+
+```tsx
+// Extend by adding a slot
+// Mutate state in extended component and call mergeProps again
+const baseState = useBaseComponent(props, ref);
+
+// React elements cannot be extended and will break `resolveShorthandProps`
+// set to undefined since it will be resolved again anyway
+(baseState as unknown) as ExtendedComponentProps).slot = undefined;
+const state = mergeProps(baseState, defaultProps, resolveShorthandProps(props, extendedComponentShorthandProps));
```

Comment:
where is `extendedComponentShorthandProps` coming from. I'm having difficulties to understand what is going on in this example